pykechain/exceptions.py

```diff
@@ -15,30 +14,54 @@ class APIError(Exception):
     :ivar detail: details of the error
     """
 
-    def __init__(self, *args, **kwargs):
+    def __init__(self, text, *args, **kwargs):
         """Initialise the `APIError` with `response`, `request`, `msg`, `traceback` and `detail`.
 
         :param response:
         :param kwargs:
         """
-        response = kwargs.pop('response', None)
-        self.response = response
-        self.request = kwargs.pop('request', None)
-        if (response is not None and not self.request and hasattr(response, 'request')):
+        self.response = kwargs.pop('response', None)
+
+        try:
             self.request = self.response.request
+        except AttributeError:
+            self.request = kwargs.pop('request', None)
+
         self.msg, self.traceback, self.detail = None, None, None
-        super(APIError, self).__init__(*args, **kwargs)
 
-        if isinstance(self.response, Response) and 'content' in self.response:
+        context = [text]
+
+        if self.response is not None and isinstance(self.response, Response):
+            response_json = self.response.json()
             # may be a KE-chain API response error
-            if self.response.json():
-                self.msg = self.response.json().get('msg')
-                self.traceback = self.response.json().get('traceback')
-                self.detail = self.response.json().get('detail')
+            if response_json:
+                self.msg = response_json.get('msg')
+                self.traceback = response_json.get('traceback')
+                self.detail = response_json.get('detail')
             else:
                 self.traceback = self.response.text
                 self.msg = self.traceback.split(r'\n'[-1])
 
+            context.extend([
+                'Server {}'.format(self.traceback),
+                'Elapsed: {}'.format(self.response.elapsed),
+            ])
+
+        if self.request is not None and isinstance(self.request, PreparedRequest):
```

Comment:
Nice this.

We need to test this however with several 'injections' because not all `ApiErrors` are created equal.

Reply:
Is there a way to produce these injections?

Reply:
some try-excepts fix known issues.

Reply:
It is always a it unknown what content the ApiErrors will receive. I guess you now cover most cases. A way is to Find all ApiErrors in the code and see with what arguments they are raised. 

Basically by checking and chanhing the signature of APIErrors and changing them all in the code you already do exactly this.